feat: ranks modal

rebelchris opened the pull request:
## Changes

- Introducing the new rank modal
- Not finished see below for missing parts

## Missing

- ~~Actual data changes from the API~~
- The new icons
- ~~Colors for each rank~~
- Tags from API
- ~~Earn text based on actual rank status~~

## Testing

- Added tests for the new ranks

DD-508 #done
DD-507 #done

ayadotdev commented:
Great job Chris! 🥇

sshanzel commented on packages/shared/src/lib/rank.ts:
```diff
@@ -2,6 +2,74 @@ export const NO_RANK = 'No rank';
 export const FINAL_RANK = 'You made it 🎉';
 export const STEPS_PER_RANK = [3, 4, 5, 6, 7];
 export const RANK_NAMES = ['Bronze', 'Silver', 'Gold', 'Platinum', 'Diamond'];
```
We probably need to update this as well or the areas that implement this to utilize the new one. What do you think?

rebelchris replied:
Yep we'll use the new RANKS object for this as well, wanted to take this on in DD-507 (Aya's ticket)

rebelchris replied:
Actually @sshanzel decided to add this in this PR :)
Feel free to check again

sshanzel reviewed (commented):
This is superb! Just had one question but this PR looks great already!

idoshamun commented on packages/shared/src/components/modals/RanksModal.tsx:
```diff
@@ -79,88 +25,183 @@ export interface RanksModalProps extends ModalProps {
+                      }-color)`,
+                    } as CSSProperties
+                  }
+                />
+              )}
+              <Rank
+                rank={_rank.level}
+                className={classNames(
+                  rank === _rank.level ? 'w-10 h-10' : 'w-8 h-8',
+                )}
+                colorByRank={finalRankCompleted ? true : showRank > _rank.level}
+              />
+            </div>
+            <strong
+              className={classNames(
+                'mt-2',
+                rank === _rank.level
+                  ? 'typo-callout text-theme-status-success'
+                  : 'typo-footnote',
+                rankCompleted(showRank, _rank, progress)
+                  ? _rank.color
+                  : 'text-theme-label-tertiary',
+              )}
+            >
+              {_rank.name}
+            </strong>
+          </li>
+        ))}
+      </ul>
+      <p className="mt-1 mb-3 text-center text-theme-label-tertiary typo-footnote">
+        {getNextRankText({
+          nextRank: rank + 1,
+          rank,
+          finalRank: isFinalRank,
+          progress,
+          showNextLevel: false,
+        })}
+      </p>
+    </section>
+  );
```
Looks somewhat lengthy. I think we can refactor a bit and split to helper components

idoshamun commented on packages/shared/src/components/modals/RanksModal.tsx:
```diff
@@ -1,73 +1,19 @@
-  ReactNode,
-  useContext,
-} from 'react';
+import React, { CSSProperties, ReactElement, useContext } from 'react';
```
Let's split this file into multiple files. It has so many things going on here. We can create a directory for the `RanksModal` and have everything there

rebelchris replied:
On it 🥳

idoshamun reviewed (changes requested):
Massive! I have some comments inside.
Make sure to test also the new rank modal and see that everything works there with these changes.

rebelchris commented:
@idoshamun I've refactored everything into small components and added some classed components to make things more clear.

Let me know what you think 🥳
@sshanzel Would also love to hear your thoughts on naming espacially.

sshanzel commented on packages/shared/src/components/modals/common.tsx:
```diff
@@ -4,3 +4,7 @@ export const ModalHeader = classed(
   'header',
   'flex flex-row items-center py-2 px-3 h-14 border-b border-theme-divider-tertiary',
```
Not sure if this is part of the new guideline - but the padding for the header is now `24px` = `px-6`. If so, then should we make small updates from here or do it in a single separate PR?
```suggestion
  'flex flex-row items-center py-3 px-3 h-14 border-b border-theme-divider-tertiary',
```

rebelchris replied:
Let's refactor right away, it should reflect all headers anyway.
I will add another ticket to revisit modals as they don't all use this sytem.

sshanzel commented on packages/shared/src/components/modals/RanksModal/TimezoneText.tsx:
```diff
@@ -0,0 +1,40 @@
+  );
+
+  const loggedTimezoneDescription = !user?.timezone ? accountDetails : null;
+  const timezoneDescription = !user ? signIn : loggedTimezoneDescription;
```
Just my personal POV - I think we can return early if the user does not exist so we can reduce the ternary operators. Just felt a bit of difficulty reading these two lines. But feel free to ignore this if it won't make too much sense.
Ex:
```
if (!user) {
  return <ModalText className={classes}>
    To fit the weekly goal to your time zone, please sign up and add it in the
    account details.
  </ModalText>
}
```

rebelchris replied:
Sure sounds great, i'll change it :)

sshanzel commented on packages/shared/src/components/modals/RanksModal/index.tsx:
```diff
@@ -0,0 +1,55 @@
+  useTrackModal({ isOpen: props.isOpen, title: 'ranks modal' });
+  const currentRank = rank;
+
+  // Todo: Refactor these to load from API
```
Just tagging for reference. API should now be available. Woohoo!

sshanzel reviewed (commented):
Looks neat! ✨ 

Just had some non-blocking comments/questions.

idoshamun commented:
@rebelchris something fails here. Just wanted to let you know while I'm reviewing

idoshamun reviewed (approved):
This looks so much better!
Clean, neat, well organized. I love it.

sshanzel reviewed (approved):
Spotless! ✨

idoshamun reviewed (approved):
This looks good, my only worry is that we not tested it enough.
This feature has so many edge cases, please do your best to cover them manually.
